### PR TITLE
ALIAS: Respond SERVFAIL on non-NOERRORs from resolver

### LINKED ISSUE #6666: ALIAS: behaviour when resolver returns anything but NOERROR
- Program: Authoritative
 - Issue type: Bug report

### Short description
At the moment, in 4.1.2 and master, the Authoritative Server will happily forward whatever information it gets in response to an ALIAS lookup. For instance, when the ALIAS target does not exist, the response to the client will be NXDOMAIN. This could effectively "blank out" a name or domain, especially if the ALIAS is at the apex of a zone.

A comment in the code for ALIAS expansion over TCP leads me to believe (and I agree with this) that we should answer with SERVFAIL in this case. So the client can retry another authoritative server.

I have a branch [here](https://github.com/pieterlexis/pdns/commits/alias-servfail-on-nxd), but the tests are failing because the SERVFAIL also (rightfully) kills the AXFR with an expanded ALIAS. I think I'll make a separate zone for this test to test this behaviour. This ticket is created so we don't forget.

### Steps to reproduce
1. create ALIAS to non-existent domain
2. Query for A record of the ALIAS name
3. Get NXDOMAIN for that name

### Expected behaviour
SERVFAIL

### Actual behaviour
NXDOMAIN

### PR DESCRIPTION
**Pull request by pieterlexis**
### Short description
In #6659 is was discovered that the Authoritative Server would send NXDOMAIN if the ALIAS target would yield NXDOMAIN (actually, it would pass the RCODE from the resolver). Leading to potential issues of blanking out a domain if the ALIAS was on the apex.

This PR fixes these issues by setting the rcode to SERVFAIL if the recursor's answer is not NOERROR. 

On timeouts from the resolver, we will send SERVFAIL as well over TCP. But due to the design of the dnsproxy there is no response over UDP, it will take some work to be able to detect this, we'll have to read our conntrack table and return SERVFAIL if some time has expired. All in a loop separate from the recvfrom() loop we already have in `DNSProxy::mainloop`.

Closes #6666

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [x] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by pieterlexis** [comment]
A yes, I know remember why this was not PR'd before. Because we SERVFAIL the AXFR when the ALIAS can not the expanded, the tests fail.

**Comment by gibson042** [comment]
> setting the rcode to SERVFAIL if the recursor's answer is not NOERROR

Isn't that overly blunt? NXDOMAIN in particular seems like it should be translated to NODATA (i.e., RCODE=NOERROR + ANCOUNT=0).

And there's probably already an ALIAS bug with respect to NODATA—negative responses must include the SOA of the containing zone.

**Comment by pieterlexis** [comment]
> Isn't that overly blunt? NXDOMAIN in particular seems like it should be translated to NODATA (i.e., RCODE=NOERROR + ANCOUNT=0).

This is something that could be argued. NXDOMAIN of the ALIAS target in my eyes means that the zone was misconfigured as the ALIAS target must exist. But I am open to be convinced differently.

> And there's probably already an ALIAS bug with respect to NODATA—negative responses must include the SOA of the containing zone.

From a quick test, this indeed seems to be the case. We'll fix that. Thanks!

**Comment by pieterlexis** [comment]
@Habbie the tests pass, could you review?

File: pdns/dnsproxy.cc [hunk 1]
```diff
@@ -100,7 +100,7 @@ bool DNSProxy::completePacket(DNSPacket *r, const DNSName& target,const DNSName&
       ret2 = stubDoResolve(target, QType::AAAA, ips);
 
     if(ret1 != RCode::NoError || ret2 != RCode::NoError) {
-      g_log<<Logger::Error<<"Error resolving for "<<aname<<" ALIAS "<<target;
+      g_log<<Logger::Error<<"Error resolving for "<<aname<<" ALIAS "<<target<<" over TCP";
```
Comment (Habbie) [inline pdns/dnsproxy.cc]: I didn't think we actually did a TCP query ourselves, here, so probably needs better wording. Given our entirely distinct handling for whether the client query came via UDP or TCP, logging the difference does seem sensible.

File: pdns/dnsproxy.cc [hunk 2]
```diff
@@ -245,26 +245,31 @@ void DNSProxy::mainloop(void)
         /* Set up iov and msgh structures. */
         memset(&msgh, 0, sizeof(struct msghdr));
         string reply; // needs to be alive at time of sendmsg!
-        if(i->second.complete) {
-
+        if(i->second.complete) { // TODO in the world of non-recursive backends, isn't this always true?
```
Comment (Habbie) [inline pdns/dnsproxy.cc]: Indeed! Nothing in our codebase calls `DNSProxy::recurseFor` anymore.

File: regression-tests.auth-py/test_ALIAS.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    def testNxDomain(self):
+        query = dns.message.make_query('nxd.example.org', 'A')
+        res = self.sendUDPQuery(query)
+        self.assertRcodeEqual(res, dns.rcode.SERVFAIL)
+
+        query = dns.message.make_query('nxd.example.org', 'AAAA')
+        res = self.sendUDPQuery(query)
+        self.assertRcodeEqual(res, dns.rcode.SERVFAIL)
+
+        # TODO this should actually return SOA + NS?
```
Comment (Habbie) [inline regression-tests.auth-py/test_ALIAS.py]: No - ANY includes A/AAAA and thus ALIAS and thus I think SERVFAIL is fine.

**Comment by Habbie** [comment]
> And there's probably already an ALIAS bug with respect to NODATA—negative responses must include the SOA of the containing zone.

The PR does not appear to fix this, but this is tracked in #4556

File: regression-tests.auth-py/test_ALIAS.py (hunk 3 above)
Comment (pieterlexis) [reply regression-tests.auth-py/test_ALIAS.py]: ANY is not ALL though....

Comment (Habbie) [reply regression-tests.auth-py/test_ALIAS.py]: > ANY is not ALL though....

Yes it is. Cloudflare can go ** a ** :)

**Pull request merged**